flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/rel/logical/LogicalTableScan.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.calcite.rel.logical;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.plan.Convention;
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptTable;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelInput;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.hint.RelHint;
+import org.apache.calcite.schema.Table;
+
+import java.util.List;
+
+/**
+ * A <code>LogicalTableScan</code> reads all the rows from a {@link RelOptTable}.
+ *
+ * <p>This class is copied from Calcite because the {@link #explainTerms} should consider hints.
+ *
+ * <p>If the table is a <code>net.sf.saffron.ext.JdbcTable</code>, then this is literally possible.
+ * But for other kinds of tables, there may be many ways to read the data from the table. For some
+ * kinds of table, it may not even be possible to read all of the rows unless some narrowing
+ * constraint is applied.
+ *
+ * <p>In the example of the <code>net.sf.saffron.ext.ReflectSchema</code> schema,
+ *
+ * <blockquote>
+ *
+ * <pre>select from fields</pre>
+ *
+ * </blockquote>
+ *
+ * <p>cannot be implemented, but
+ *
+ * <blockquote>
+ *
+ * <pre>select from fields as f
+ * where f.getClass().getName().equals("java.lang.String")</pre>
+ *
+ * </blockquote>
+ *
+ * <p>can. It is the optimizer's responsibility to find these ways, by applying transformation
+ * rules.
+ *
+ * <p>CALCITE-4581, Line 106: {@link #explainTerms} method should consider hints.
```

Comment:
Let's use the new strategy to trace modification? i.e. `// BEGIN FLINK MODIFICATION`